### PR TITLE
Fixed concurrent modification on reload subscriptions from DB

### PR DESCRIPTION
**Pull request by abeykoon**
Fixed concurrent modification on reload subscriptions from DB. Also added necessary used locks.

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java [hunk 1]
```diff
@@ -39,6 +41,9 @@
 
     private List<SubscriptionListener> subscriptionListeners = new ArrayList<>();
 
+    private final ReadWriteLock clusterSubscriptionModifyLock = new ReentrantReadWriteLock();
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: Document this variable.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java [hunk 2]
```diff
@@ -39,6 +41,9 @@
 
     private List<SubscriptionListener> subscriptionListeners = new ArrayList<>();
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: document this one as well if we have not.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java (hunk 1 above)
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: and explain why we need a lock etc.

Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: create a jira and link it to this modification.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java [hunk 3]
```diff
@@ -322,91 +336,123 @@ public void updateClusterSubscriptionMaps(AndesSubscription subscription, Subscr
...
-            while (memSubIterator.hasNext()) {
-                AndesSubscription currentSubscription = memSubIterator.next();
-                if (!dbSubscriptions.contains(currentSubscription)) {
-                    log.warn("Cluster Subscriptions are not in sync. Subscriptions exist in memory that are not " +
-                            "available in db. Thus removing from memory " + currentSubscription);
-                    subscriptionStore.createDisconnectOrRemoveClusterSubscription(currentSubscription, SubscriptionListener
-                            .SubscriptionChange.DELETED);
+                //check if each memory subscription is in DB. If not remove from memory
+                if(null != memorySubscriptions) {
+                    Iterator<AndesSubscription> memSubIterator = memorySubscriptions.iterator();
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: can't we do a for (....) here ? given that we have change the way we do things.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java [hunk 4]
```diff
@@ -322,91 +336,123 @@ public void updateClusterSubscriptionMaps(AndesSubscription subscription, Subscr
...
-            while (memSubIterator.hasNext()) {
-                AndesSubscription currentSubscription = memSubIterator.next();
-                if (!dbSubscriptions.contains(currentSubscription)) {
-                    log.warn("Cluster Subscriptions are not in sync. Subscriptions exist in memory that are not " +
-                            "available in db. Thus removing from memory " + currentSubscription);
-                    subscriptionStore.createDisconnectOrRemoveClusterSubscription(currentSubscription, SubscriptionListener
-                            .SubscriptionChange.DELETED);
-                    notifyClusterSubscriptionHasChanged(currentSubscription, SubscriptionListener.SubscriptionChange.DELETED);
+                //check if each memory subscription is in DB. If not remove from memory
+                if(null != memorySubscriptions) {
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: can't we do for (...) loop? ( will bring better clarity ... my personal opinion)

Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: Agreed. Updated with modified PR.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java (hunk 3 above)
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: Agreed. Updated with modified PR.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java (hunk 1 above)
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: Added Jira https://wso2.org/jira/browse/MB-1213

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java (hunk 2 above)
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: Fixed with modified PR.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java [hunk 5]
```diff
@@ -146,28 +151,33 @@ public void addSubscription(LocalSubscription localSubscription) throws AndesExc
...
-                    mockSubscription.close();
+                        LocalSubscription mockSubscription = convertClusterSubscriptionToMockLocalSubscription(sub);
+                        mockSubscription.close();
 
-                    /**
-                     * Close and notify. This is like closing local subscribers of that node thus we need to notify
-                     * to cluster.
-                     */
-                    LocalSubscription removedSubscription = subscriptionStore.removeLocalSubscription(mockSubscription);
+                        /**
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesSubscriptionManager.java]: why would you put a java-doc comment inside a method?

**Pull request merged**